task/jenkins 37715 npm shrinkwrap

cliffmeyers opened the pull request:
# Description
- This specifically adds npm-shrinkwrap to `blueocean-dashboard`, `blueocean-personalization`, `blueocean-web` and `blueocean-config` 
- See [JENKINS-37715](https://issues.jenkins-ci.org/browse/JENKINS-37715).

On my machine, current master, no shrinkwrap,
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Blue Ocean UI Parent ............................... SUCCESS [  5.277 s]
[INFO] Module :: BlueOcean :: Commons API ................. SUCCESS [  5.833 s]
[INFO] Module :: BlueOcean :: Web module .................. SUCCESS [02:44 min]
[INFO] Module :: BlueOcean :: Rest module ................. SUCCESS [  4.665 s]
[INFO] Module :: BlueOcean :: JWT module .................. SUCCESS [ 31.002 s]
[INFO] Module :: BlueOcean :: REST API implementation ..... SUCCESS [01:22 min]
[INFO] Module :: BlueOcean :: Pipeline REST API implementation SUCCESS [02:23 min]
[INFO] Module :: BlueOcean :: Events ...................... SUCCESS [  5.819 s]
[INFO] Module :: BlueOcean :: Dashboard ................... SUCCESS [03:16 min]
[INFO] Module :: BlueOcean :: Personalization ............. SUCCESS [03:19 min]
[INFO] Module :: BlueOcean :: Config ...................... SUCCESS [01:54 min]
[INFO] BlueOcean beta ..................................... SUCCESS [ 25.109 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 16:21 min
[INFO] Finished at: 2016-09-12T17:46:24-04:00
[INFO] Final Memory: 102M/1553M
[INFO] ------------------------------------------------------------------------
```

this branch, after shrinkwrap
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Blue Ocean UI Parent ............................... SUCCESS [  4.752 s]
[INFO] Module :: BlueOcean :: Commons API ................. SUCCESS [  5.668 s]
[INFO] Module :: BlueOcean :: Web module .................. SUCCESS [01:23 min]
[INFO] Module :: BlueOcean :: Rest module ................. SUCCESS [  4.741 s]
[INFO] Module :: BlueOcean :: JWT module .................. SUCCESS [ 31.159 s]
[INFO] Module :: BlueOcean :: REST API implementation ..... SUCCESS [01:22 min]
[INFO] Module :: BlueOcean :: Pipeline REST API implementation SUCCESS [02:19 min]
[INFO] Module :: BlueOcean :: Events ...................... SUCCESS [  5.964 s]
[INFO] Module :: BlueOcean :: Dashboard ................... SUCCESS [01:34 min]
[INFO] Module :: BlueOcean :: Personalization ............. SUCCESS [01:24 min]
[INFO] Module :: BlueOcean :: Config ...................... SUCCESS [ 42.362 s]
[INFO] BlueOcean beta ..................................... SUCCESS [ 24.431 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10:05 min
[INFO] Finished at: 2016-09-12T17:57:54-04:00
[INFO] Final Memory: 101M/950M
[INFO] ------------------------------------------------------------------------
```

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

michaelneale commented:
woah nice - needs some updating. 

@sophistifunk @scherler you want to take a look at this and see if you like?

i386 commented:
Whoa 6 minutes?

cliffmeyers commented:
👍

sophistifunk commented:
I think everybody knows I'm pro-shrinkwrap :) 🐝

scherler commented on blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx:
```diff
@@ -31,8 +31,9 @@ const QueuedState = () => (
                 icon: 'timer',
                 style: { fill: '#fff' },
-            }} />
-            <span className="waiting">Waiting for run to start.</span>
```
That will make AT fail!!! :bug:

sophistifunk replied:
Why?

cliffmeyers replied:
This is just a whitespace change. We had a peer dep error for months because the jenkins-es-lint-config plugin required ^4.0.1 and dashboard was using 5.x. You can't shrinkwrap when you have peer dep errors so the easiest fix was to roll back to 4.0.1 and make a handful of whitespace changes.

scherler replied:
it will fail because you remove the ```@className```

cliffmeyers replied:
Very strange, that must have been a mismerge. Definitely didn't delete it intentionally. Will fix

cliffmeyers replied:
Fixed in 00c3709

cliffmeyers commented:
@schlerer basically once the shrinkwrap. json already exists you just use "npm install --save" or --saveDev when you need to make changes to deps. It will update package.json and shrinkwrap.json for you. I can add a note in the readme.

see: https://docs.npmjs.com/cli/shrinkwrap#building-shrinkwrapped-packages

cliffmeyers commented:
Just to shed some light on the overall process here for those interested. For each module:
- Cleaned node_modules and ran a fresh "npm install"
- Noted the version number for all devDependencies
- Aligned version numbers across all modules -  these were consistent in pretty much all cases except for one or two things like enzyme
- Updated package.json to use exact version numbers just installed
- Ran "npm shrinkwrap"
